editor/utils/dom.js

```diff
@@ -1,3 +1,50 @@
+const NODE_TEXT_TYPE = 3;
+const NODE_ELEMENT_TYPE = 1;
+
+function getCursorStart( /* node */ ) {
+	// NOTE: In the future, we may want to skip some non-inhabitable positions in some nodes
+	return 0;
+}
+
+function getCursorEnd( node ) {
+	// NOTE: In the future, this might be a place to ignore special characters that browsers ignore
+	if ( node.nodeType === NODE_TEXT_TYPE ) {
+		return node.nodeValue.length;
+	}
+
+	// We may need exceptions for other empty tags as well (e.g. hr, br, input)
+	if ( node.nodeType === NODE_ELEMENT_TYPE && node.nodeName.toLowerCase() === 'img' ) {
+		return 1;
+	}
+
+	// Non text nodes have a last cursor position of their number of children
+	return node.childNodes.length;
+}
+
+/**
+ * Check whether the offset is at the start of the node
+ *
+ * @param {Element} node    the DOM element
+ * @param {Integer} offset  the offset
+ * @return {Boolean}        whether or not the offset is at the first cursor position in node
+ */
+export function isAtCursorStart( node, offset ) {
+	const nodeStart = getCursorStart( node );
+	return nodeStart === offset;
+}
+
+/**
+ * Check whether the offset is at the end of the node
+ *
+ * @param {Element} node    the DOM element
+ * @param {Integer} offset  the offset
+ * @return {Boolean}        whether or not the offset is at the last cursor position in node
+ */
+export function isAtCursorEnd( node, offset ) {
```

Comment:
Might it make sense to only introduce one method (`isAtCursorEnd`) for now? In more sure if it's useful to create 4 functions right now. Additionally, if editable uses the rest of the logic in this file, we might not need to export it.

Reply:
You would, for the tests, but maybe we can test `isEdge` as a whole instead?

Reply:
With your above change to just use isEdge outside instead of needing to call an internal method, my preference would be to keep the methods for isAtCursorStart, but only expose isEdge. How does that sound?

Reply:
The major reason is it just increases the likelihood that lots of code won't have to change if it turns out that you have to start ignoring some offsets for start (like empty text nodes). It can get fairly gnarly with text node fragmentation. Having the cursor calculations in one spot helps in the long run. But yep, if we are able to use isEdge the other place I changed, then definitely only expose that one and just test it.

Reply:
> With your above change to just use isEdge outside instead of needing to call an internal method, my preference would be to keep the methods for isAtCursorStart, but only expose isEdge. How does that sound?

That's exactly what I mean.

Reply:
Only isEdge is exposed now.

Reply:
What I also meant though: maybe we can only keep only `getCursorEnd ` if the other one is just a stub for now, and `isAtCursorEnd` is just an equality check? Why create so many functions?